### PR TITLE
#2991: keep references to bound parameter values in sasql driver

### PR DESCRIPTION
**Pull request by andipohl**

### CODE REVIEW

**Comment by Ocramius** [comment]
@andipohl this needs a test case as well

**Comment by andipohl** [comment]
I created a testcase, but travis seems to be failed because internal server problems

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 use Doctrine\DBAL\Driver\SQLAnywhere\Driver;
 use Doctrine\DBAL\DriverManager;
+use PDO;
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: Is this needed?

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php [hunk 2]
```diff
@@ -50,4 +51,54 @@ public function testPersistentStatement()
...
+     */
+    public function testBindParams()
+    {
+        $params = $this->_conn->getParams();
+
+        $conn = DriverManager::getConnection($params);
+
+        $conn->connect();
+        $conn->beginTransaction();
+        try {
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: I'd avoid having flow control structures in tests as much as possible. All side effects should be handled in `tearDown()`.

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php [hunk 3]
```diff
@@ -50,4 +51,54 @@ public function testPersistentStatement()
...
+            self::assertTrue($conn->isConnected(),'No SQLAnywhere-Connection established');
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: As far as I can tell, the failure to connect will throw an exception, so this line of code will not be reached. In any event, this particular test shouldn't check connection since it's not about it.

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php [hunk 4]
```diff
@@ -50,4 +51,54 @@ public function testPersistentStatement()
...
+
+            $sql = 'DROP TABLE IF EXISTS testTable2991';
+            $prepStmt = $conn->prepare($sql);
+            $prepStmt->execute();
+
+            $sql = 'CREATE TABLE IF NOT EXISTS testTable2991(
+                      col1 VARCHAR(80 CHAR) NOT NULL,
+                      col2 VARCHAR(20) NOT NULL,
+                      col3 INTEGER NOT NULL
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: Are three columns required to reproduce the case or fewer could be used?

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php [hunk 5]
```diff
@@ -50,4 +51,54 @@ public function testPersistentStatement()
         self::assertTrue($prepStmt->execute(),' Statement persistent failed');
     }
 
+    /**
+     * Check Issue #2991
+     * @throws \Doctrine\DBAL\ConnectionException
+     * @throws \Doctrine\DBAL\DBALException
+     * @throws \Exception
+     */
+    public function testBindParams()
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: The case itself looks rather generic and not depending on a specific platform. Could it be moved to `Doctrine\Tests\DBAL\Functional\StatementTest`?

Besides that, the test case above already contains a tests which covers the `prepare()` → `bindValue()` → `execute()` flow (see `testReuseStatementWithReboundValue()`). Is this case somehow different from that one?

**Review by morozov — CHANGES_REQUESTED** [review]
@andipohl I've restarted the build. Please see [code style violations](https://travis-ci.org/doctrine/dbal/jobs/333214553).

File: lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php [hunk 6]
```diff
@@ -63,6 +63,15 @@ class SQLAnywhereStatement implements IteratorAggregate, Statement
      */
     private $stmt;
 
+    /**
+     * Holds references to bound parameter values.
+     *
+     * This is a new requirement for PHP7's oci8 extension that prevents bound values from being garbage collected.
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php]: The "oci8" piece looks irrelevant.

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php (hunk 1 above)
Comment (andipohl) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: corrected

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php (hunk 2 above)
Comment (andipohl) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: removed and createTable/dropTable moved to setUp/tearDown

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php (hunk 3 above)
Comment (andipohl) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: removed

File: lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php (hunk 6 above)
Comment (andipohl) [reply lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php]: sorry, typo error

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php (hunk 4 above)
Comment (andipohl) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: I reduced it to 2 column, a varchar and an int. Only with this preparation the error
```
Caused by
Doctrine\DBAL\Driver\SQLAnywhere\SQLAnywhereException: SQLSTATE [00000] [-158] Value 139645189756408 out of range for destination
```
can be reproduced.
The driver access memory, which is freed

File: tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php (hunk 5 above)
Comment (andipohl) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php]: ok, I moved it. Maybe another driver has the same problem, like sasql and furthermore oci8.
testReuseStatementWithReboundValue() can not reproduced the error. I think, we need a string-parameter.

File: tests/Doctrine/Tests/DBAL/Functional/StatementTest.php [hunk 7]
```diff
@@ -186,6 +186,27 @@ public function testReuseStatementWithReboundValue()
         self::assertEquals(2, $stmt->fetchColumn());
     }
 
+    /**
+     * Issue #2991
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/StatementTest.php]: Please use `@group DBAL-2991` instead.

**Review by morozov — APPROVED** [review]
Looks good to me.

File: lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php [hunk 8]
```diff
@@ -110,6 +119,8 @@ public function bindParam($column, &$variable, $type = null, $length = null)
                 throw new SQLAnywhereException('Unknown type: ' . $type);
         }
 
+        $this->boundValues[$column] =& $variable;
```
Comment (Majkl578) [inline lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php]: This should probably be `= &$variable`, but we have no CS rule for this afaik so it's a minor.

File: tests/Doctrine/Tests/DBAL/Functional/StatementTest.php (hunk 7 above)
Comment (andipohl) [reply tests/Doctrine/Tests/DBAL/Functional/StatementTest.php]: done

File: lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php (hunk 8 above)
Comment (andipohl) [reply lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php]: done. But I made copy&paste from oci8-Statement ;-)

File: tests/Doctrine/Tests/DBAL/Functional/StatementTest.php [hunk 9]
```diff
@@ -186,6 +186,27 @@ public function testReuseStatementWithReboundValue()
...
+    public function testBindParamsWithMemoryCorruption()
+    {
+        $sql = 'INSERT INTO stmt_test(id, name) VALUES (?, ?)';
+
+        $col1 = 0;
+        $col2 = '20';
+
+        $stmt = $this->_conn->prepare($sql);
+
+        $paramCount = 1;
```
Comment (Ocramius) [inline tests/Doctrine/Tests/DBAL/Functional/StatementTest.php]: Why is the `++` approach used here? :|

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by Ocramius** [comment]
@andipohl make sure you mail me directly about this one.

**Comment by morozov** [comment]
Closing as SQLAnywhere is no longer supported.
